File: iter1/❌560.cpp
// preSum + map
//

#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int sum = 0, result = 0;
        unordered_map<int, int> preSum;
        preSum[0] = 1;
        for (auto n : nums) {
            sum += n;
            if (preSum.find(sum - k) != preSum.end()) {
                result += preSum[sum-k];
            }
            if (preSum.find(sum) != preSum.end()) {
                preSum[sum]++;
            } else {
                preSum[sum] = 1;
            }
        }
        return result;
   }
};

int main() {
    Solution s;
    vector<int> v{1,1,1};
    cout << s.subarraySum(v, 2) << endl;
}

File: iter2/❌560.py
# prefix_sum
# why can it work? because it uses hashmap to help
# find the previous valid positions in O(1) time
# instead of brute force looking for every subarray.

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        pre_sum = collections.defaultdict(int)
        sum = ans = 0
        for i, num in enumerate(nums):
            sum += num
            ans += pre_sum[k-sum]
            pre_sum[sum] += 1
        return ans









